### ISSUE TITLE
AzureBlobCache - Option to issue 302 redirect to cached image in cloud or serve image directly

### ISSUE THREAD

**Issue by gnpretorius** [JimBobSquarePants/ImageProcessor#178]
The current AzureBlobCache issues a 302 redirect to the actual cached resource sitting in the cloud. This works fine in some situations and also helps high volume sites as your server only issues a 302 and does not actually need to stream the file to the browser. 

However, in some situations this is not ideal. What is needed is an option to stream the image from Azure Storage via the server. So if I request http://www.site.com/image.jpg?width=50, it serves the actual image rather than the 302. The purpose of this setting is for CDN environments. 

Azure CDN now supports custom origins, meaning you can specify your site as the origin for all the content on the CDN. This means that any request done via the CDN will check your site for the content, cache it and then serve it. Any subsequent request will be cached at the CDN level thus reducing the amount of requests on your server drastically. Currently the sites issues a 302, which the CDN will not cache meaning every request will hit your server. 

I've just finished a site serving 30GB of images an hour and this CDN has a massive impact on the performance of the site. We used http://www.imageresizer.net/ but this is limited as it does not support caching in the cloud and comes at a price. 

We are doing another similar large scale project and I would love to use this library. 

Will definitely donate some beers :)

G

**Comment by cosmo0** [comment]
Related to #161 ?

> However, in some situations this is not ideal

Could you explain which situations? :)

**Comment by gnpretorius** [comment]
No problem. 

It's when you have a CDN running and your site is set as the origin source for the CDN. i.e. if someone requests an image, it should be served from the site (not a 302) so that the CDN can cache the image. With a 302 it's not cached. 

Request --> CDN --> Your server --> Image
Next requests
Request --> CDN
Request --> CDN
Request --> CDN

Let me know if this makes sense. 

Thanks

G

**Comment by cosmo0** [comment]
So, if your site is set as the origin source for the CDN, it's the bandwidth of your server that serves to transport the image data ?  
If the client is always hitting your server and the CDN is sending data through your server, how does the client knows to hit the CDN directly in subsequent queries ? :confused:

**Comment by gnpretorius** [comment]
Let's say you have 2 domains,

http://www.site.com
http://cdn.site.com (This is the CDN site with it custom origin setup as http://www.site.com)

When accessing the site, users will use the first domain. All images, css and javascript is referenced using second domain.

When the page loads, the request checks the CDN. There is a 99.9% chance that the CDN has the resource you are looking for and serves the cached version. If it does not, it then fetches the resource from the first domain and caches it within the CDN infrastructure. 

This means that only a small number of request will hit your server, the rest hitting the CDN and being served from there. 

With the current 302 redirect, the CDN never caches the request meaning all request are forwarded to your server and you do not benefit from the global cache distribution. 

Hopefully that makes sense :smiley: 

G

**Comment by cosmo0** [comment]
Ah ! I think I understand it better. I thought that you were setting things up the other way, that your images were referenced on www.site.com and that your server was redirecting the requests to the CDN. It makes much more sense if they are directly pointing on the CDN :smile:

**Comment by gnpretorius** [comment]
Excellent!

I was looking to modify the ```AzureBlobCache``` file to support this but not sure the interface supports this. 

It calls ```CreateCachedFileNameAsync``` and then ```RewritePath```. Would I implement this in ```RewritePath```?

Can I just serve the image from the cloud straight to this context?

Thanks again!

G

**Comment by JimBobSquarePants** [comment]
Man I find this hard to follow.... 

How do you ensure that the request reaches ImageProcessor at all? 

- What url is sent from the CDN to your server?
- What is generating the original request url? (A server without ImageProcessor?) At the moment you'd need to know the resultant cached path in order to correctly check if the image exists on the CDN.

e.g. `http://yourcdn.windows.net/cache/b/2/3/0/d/9/b230d9c0081e39a65758b8f2a9c427fa21344a2a.jpg `

This url is automatically generated from your cached blob file (ImageProcessor has no power over your CDN url)

If generating the correct url is possible. you'd have to handle your changes in `RewritePath`. I can't change the interface for obvious reasons. You might have to tweak the HttpModule also.

When submitting a pull requests could you document the pipeline in the description also to make it easier for myself and other developers to follow.

**Comment by JimBobSquarePants** [comment]
Added in version [1.0.4](https://www.nuget.org/packages/ImageProcessor.Web.Plugins.AzureBlobCache/1.0.4) 

Thanks everyone for your persistence and contribution. It is is very much appreciated :+1:
